format.js

```diff
@@ -0,0 +1,327 @@
+  }
+  const msgVal = fromNativeMsg(nativeMsg, encoding)
+  const content = bipf.decode(plaintextBuf)
+  msgVal.content = content
+  return msgVal
+}
+
+function _toNativeMsgJS(msgVal) {
+  const authorBFE = bfe.encode(msgVal.author)
+  const parentBFE = bfe.encode(msgVal.parent)
+  const sequence = msgVal.sequence
+  const timestamp = msgVal.timestamp
+  const previousBFE = bfe.encode(msgVal.previous)
+  const tag = msgVal.tag
+  const contentBuffer = bipf.allocAndEncode(msgVal.content)
+  const contentHash = msgVal.contentHash
+  const value = [
+    authorBFE,
+    parentBFE,
+    sequence,
+    timestamp,
+    previousBFE,
+    tag,
+    contentBuffer.length,
+    contentHash,
+  ]
+  const encodedValue = bipf.allocAndEncode(value)
+  const signature = msgVal.signature
+  return bipf.allocAndEncode([encodedValue, signature, contentBuffer])
+}
+
+function _toNativeMsgBIPF(buffer) {
+  let authorBFE, parentBFE, sequence, timestamp, previousBFE
+  let tagBuffer, contentBuffer, contentLen, contentHash, sigBuf
+
+  const tag = varint.decode(buffer, 0)
+  const len = tag >> BIPF_TAG_SIZE
+
+  for (var c = varint.decode.bytes; c < len; ) {
+    const keyStart = c
+    var keyTag = varint.decode(buffer, keyStart)
+    c += varint.decode.bytes
+    c += keyTag >> BIPF_TAG_SIZE
+    const valueStart = c
+    const valueTag = varint.decode(buffer, valueStart)
+    const valueLen = varint.decode.bytes + (valueTag >> BIPF_TAG_SIZE)
+
+    const key = bipf.decode(buffer, keyStart)
+    if (key === 'author')
+      authorBFE = bfe.encode(bipf.decode(buffer, valueStart))
+    else if (key === 'parent')
+      parentBFE = bfe.encode(bipf.decode(buffer, valueStart))
+    else if (key === 'sequence') sequence = bipf.decode(buffer, valueStart)
+    else if (key === 'timestamp') timestamp = bipf.decode(buffer, valueStart)
+    else if (key === 'previous')
+      previousBFE = bfe.encode(bipf.decode(buffer, valueStart))
+    else if (key === 'tag') tagBuffer = bipf.decode(buffer, valueStart)
+    else if (key === 'content') {
+      if ((valueTag & BIPF_TAG_MASK) === BIPF_STRING_TYPE) {
+        contentBuffer = bipf.decode(buffer, valueStart)
+        contentLen = _base64ToBuffer(contentBuffer).length
+      } else {
+        contentBuffer = bipf.pluck(buffer, valueStart)
+        contentLen = contentBuffer.length
+      }
+    } else if (key === 'contentHash')
+      contentHash = bipf.decode(buffer, valueStart)
+    else if (key === 'signature') sigBuf = bipf.decode(buffer, valueStart)
+
+    c += valueLen
+  }
+
+  const value = [
+    authorBFE,
+    parentBFE,
+    sequence,
+    timestamp,
+    previousBFE,
+    tagBuffer,
+    contentLen,
+    contentHash,
+  ]
+  const encodedValue = bipf.allocAndEncode(value)
+  return bipf.allocAndEncode([encodedValue, sigBuf, contentBuffer])
+}
+
+function toNativeMsg(msgVal, encoding = 'js') {
+  if (encoding === 'js') {
+    return _toNativeMsgJS(msgVal)
```

Comment:
Similar here, `_toNativeFromMsgJS` and `_toNativeFromMsgBIPF`

Reply:
already done :sunglasses: